### ISSUE TITLE
Parsing where clauses

### ISSUE THREAD

**Issue by joneshf** [purescript/purescript#3297]
Something I noticed when working on `purty` is that `where` clauses are parsed and desugared at the same time: https://github.com/purescript/purescript/blob/80cef847264f4124ca20cce01e2657998bcaaeaa/src/Language/PureScript/Parser/Declarations.hs#L72-L81

Did some digging around and didn't find any strong decision for choosing to desugar as well as parse. The original PR didn't mention anything about it, anyway: https://github.com/purescript/purescript/pull/318. I did see some discussion about desugaring and parsing happening at the same time: https://github.com/purescript/purescript/issues/3023.

How do we feel about adding an explicit case to `Expr` for `where` clauses? Is it something we tried before, but didn't like? Has it not been a big concern, so nobody has done anything about it? It would help immensely in `purty`, but recognize that `purty`s case is kind of specific.

**Comment by garyb** [comment]
I'd be fine with introducing an explicit `where` constructor, we could perhaps change it's behaviour to match Haskell then too. It was never mentioned in the original PR as I barely knew what I was doing back then :wink: and yeah it hasn't really come up since.

**Comment by kRITZCREEK** [comment]
There are a few more of these things that get desugared in the parser. I'd like these to also have a place in the AST and be desugared during desugaring.

- Record puns
- Multi-argument-lambdas

I can't think of more right now, but I think there were a few, I'll add to the list if I can think of them.

**Comment by joneshf** [comment]
Cool! What's a good way to approach this? I was thinking:

* Add an additional case  `Expr` like the `Let` case (since it's no different from `Let` currently)

    ```Haskell
    data Expr
        = ...
        | Let [Declaration] Expr
        | Where [Declaration] Expr
        ...
    ```

* Change parsing to make `Where` instead of `Let`: https://github.com/purescript/purescript/blob/662da01c9f73873fcc2c4577bbc58fcc3f776f41/src/Language/PureScript/Parser/Declarations.hs#L72-L81

* Add a desugaring pass before let-binding desugaring: https://github.com/purescript/purescript/blob/662da01c9f73873fcc2c4577bbc58fcc3f776f41/src/Language/PureScript/Sugar.hs#L55-L73

That seem alright?

Some questions though. There are quite a few functions that operate on `Expr`, but pass through some cases. How do I know whether a particular function should be handling the `Where` case? I assume the traversals all have to, but what about other functions? Is it right to assume that most functions are called after desugaring? If not, how do we ensure the cases are handled appropriately? Could default to mimicking all handling of `Let` (since it's only sugar at this point). Thoughts?

**Comment by paf31** [comment]
`where` can only appear at the top of a declaration, so I think it would be better to reflect that in the types if possible.

**Comment by joneshf** [comment]
What does that mean, exactly?

There are three declarations that have an `Expr` (near as I can tell): `BindingGroupDeclaration`, `BoundValueDeclaration`, and `ValueDeclaration`

https://github.com/purescript/purescript/blob/80cef847264f4124ca20cce01e2657998bcaaeaa/src/Language/PureScript/AST/Declarations.hs#L491-L501

Should each of those be changed to something that might have a `where` clause attached to it?

Based on `parseValueWithWhereClause`: https://github.com/purescript/purescript/blob/662da01c9f73873fcc2c4577bbc58fcc3f776f41/src/Language/PureScript/Parser/Declarations.hs#L72-L81

It seems like only `BoundValueDeclaration`: https://github.com/purescript/purescript/blob/662da01c9f73873fcc2c4577bbc58fcc3f776f41/src/Language/PureScript/Parser/Declarations.hs#L99-L109 and `ValueDeclaration`: https://github.com/purescript/purescript/blob/662da01c9f73873fcc2c4577bbc58fcc3f776f41/src/Language/PureScript/Parser/Declarations.hs#L83-L91 would need to be changed. Is that right?

If so, what's a good way to change it? Should they be
```Haskell
data Declaration
  = ...
  | ValueDeclaration {-# UNPACK #-} !(ValueDeclarationData [(GuardedExpr, [Declaration])])
  | BoundValueDeclaration SourceAnn Binder Expr [Declaration]
```
?

**Comment by joneshf** [comment]
Also, as far as `purty` is concerned, the validity of a `Module` is out of scope. If you somehow write a record accessor that has a let binding with a typeclass definition in it, it'll try to pretty print it. Which is to say, it doesn't matter if it's given a `Module` with a where clause in the wrong place. It should just print it and get out of the way.

It's not up to `purty` to decide what the compiler thinks is a valid module. Main reason being, I (@joneshf) don't know all the forms a valid module can take.

**Comment by paf31** [comment]
> Also, as far as purty is concerned, the validity of a Module is out of scope.

But you're talking about changing the compiler code, where it's definitely _not_ out of scope :)

If I remember right `BindingGroupDeclaration` doesn't get created until the desugaring step, so yes it should just be the other two.

**Comment by joneshf** [comment]
I started doing this and so far have touched the `AST`, `CoreFn`, `Linter`, `Parser`, `Pretty`, `Sugar`, and `TypeChecker` hierarchies. This seems like a much bigger change than I was hoping for. Most of the places it touches, I'm not sure what to do with.

Am I doing this wrong, or is it intended to touch so many parts of the code?

**Comment by joneshf** [comment]
### Maybe let me communicate some of my concerns

* Lots of the code this is touching seems integral to the compiler working properly. I don't want to go shaking around bits that needn't be shook and create bugs somewhere unrelated to this change.
* If someone looks back on this change to figure out what happened, and they see 15+ files changed across eight hierarchies, are they going to understand why all those changes were necessary?
* When someone reviews this, are they going to remember all of the invariants that are being touched? We're all human. It's easy for us to miss something small.

There are more concerns, but we don't need a wall of them in this comment.

### And let me communicate what would be ideal

I would have liked to come in, seen a sugar AST and a real AST, added `where` clauses to the sugar AST, added a desugar step to remove `where` clauses, and made a PR. Is this a reasonable expectation? Would we be up for doing that as a first step, and then tackling this PR? Is _that_ too big of a change?

**Comment by joneshf** [comment]
Any thoughts?

**Comment by paf31** [comment]
What about annotating the `Let` constructor with a piece of data which tells you if it came from a `where` or not? It wouldn't capture the invariant that `where` only appears at the top, but it should be enough to pretty print things, and should hopefully be a lot simpler.

**Comment by garyb** [comment]
Resolved (pretty much) by #3317
